Verifier now accepts multiple errors (fixes #387).

6A opened the pull request:
Tests all pass successfully, and have been updated accordingly.

Besides the following change, everything behaves like before, meaning that every single error is reported using `return fatal!(...)`, which is equivalent to the previous `return err!(...)`.

However, from now on we can define errors as non fatal, and keep verifying. For example, instead of reporting the first error we encounter when ensuring a function's parameters are valid, we can report all errors, and stop verifying the function after checking the parameters.

The mentioned change is that now, error reporting in `pretty_error` reports **every single** reported error, instead of the first one.

I'll probably work on making most errors `nonfatal` after this.

**Edit**: I did not change the version... Maybe I should have?

**Edit 2**: Here are more details about the changes:

1. Added `VerifierErrors` type, which is a just a wrapper around a `Vec<VerifierError>`. It can also be converted to `CodegenError` easily.
2. Now, every function that returns a `VerifierResult` also takes a `&mut VerifierErrors` as parameter. Errors can be added via the macros `report!(errors, "{}", args)`, `fatal!(errors, "{}", args)` (calls `report!` and then returns `Err`), and `nonfatal!(errors, "{}", args)` (calls `report!` and returns `Ok`).
3. `VerifierResult<T>` is now defined as `Result<T, ()>`, where `Ok` represents a lack or error or an error that does not stop the verification process, and `Err` represents an error that does stop the verification process (therefore using `?` can shortcurcuit a `verify_foo` function).
4. The macro `ok_if_no_error` returns `Ok(())` if no error was reported in a `VerifierErrors`, and `Err(())` otherwise. Can be useful for methods that do not shortcircuit, but stop the verification process if they fail.
5. The macro `verify!` takes a function (ie `verify_function`, `ctx.verify`), and an arbitrary number of arguments, and will take care of creating the `VerifierErrors`, passing it to the `verify` function along with the other arguments, and will either return `Ok` if no error was reported, or `Err(VerifierErrors)` otherwise.

sunfishcode commented on lib/codegen/src/regalloc/context.rs:
```diff
@@ -134,21 +162,48 @@ impl Context {
+            }
+        }
+
+        if errors.0.is_empty() {
```
Can this condition ever be false? It seems like if we have errors, one of the `if !ok` checks above should have tripped. Can this be an assert instead?

6A replied:
It will be `false` if none of the reported errors were fatal. `ok` is `true` iff no fatal error was reported; non-fatal may have been reported, which would make the verification process stop now.

sunfishcode commented on src/wasm.rs:
```diff
@@ -148,9 +149,12 @@ fn handle_module(
+            let mut errors = VerifierErrors::default();
+            let _ = context.verify(fisa, &mut errors);
+
+            if !errors.0.is_empty() {
```
Instead of testing `errors.0.is_empty()` here, it feels cleaner to test whether the return value of `verify` `.is_err()`.

6A replied:
Ditto, the `Result` may be `Ok` but a (non-fatal) error *might* have been logged anyway.

sunfishcode commented on lib/frontend/src/lib.rs:
```diff
@@ -118,11 +118,11 @@
-//!     match res {
-//!         Ok(_) => {}
-//!         Err(err) => panic!("{}", err),
+//!     if !errors.0.is_empty() {
```
Instead of testing errors.0.is_empty() here, it feels cleaner to test whether the return value of `verify_function` .is_err().

6A replied:
Ditto. I understand that it feels a little hacky to allow an `Ok` return value without it being necessarily fine, but I thought it worked well with `?`. Furthermore, right now my objective was to introduce the new VerifierErrors type, and to use it everywhere. For now I consider (most) errors as fatal, even though some of them shouldn't halt the process completely.

sunfishcode commented on lib/codegen/src/verifier/mod.rs:
```diff
@@ -82,23 +82,77 @@ pub use self::cssa::verify_cssa;
+        } else {
+            Err(errors)
+        }
+    });
```
Could this be simplified to something like
```rust
$verif.$fun( $( $arg ),+, &mut errors).map_err(|_| errors)
```
?

6A replied:
Indeed, I'm still paranoid about using `.map_err` in macros when raw syntax could be used, but I suppose the compiler will inline it correctly.

6A replied:
Nevermind, I actually want to return an error if an error was reported (`$verif.$fun` might return `Ok` and report errors anyway).

sunfishcode commented on lib/codegen/src/verifier/mod.rs:
```diff
@@ -82,23 +82,77 @@ pub use self::cssa::verify_cssa;
+            Ok(())
+        } else {
+            Err(errors)
+        }
```
Ditto :-).

sunfishcode commented:
To answer your question above, we don't need to bump versions manually; cranelift's API isn't stable yet so we're typically bumping the version on every release right now.

sunfishcode commented on lib/codegen/src/verifier/mod.rs:
```diff
@@ -120,15 +174,58 @@ impl Display for VerifierError {
 
 /// Verifier result.
-pub type VerifierResult<T> = Result<T, VerifierError>;
+pub type VerifierResult<T> = Result<T, ()>;
```
Thought experiment: What if we split `VerifierResult` into two, like
```rust
type VerifierResult<T> = Result<T, VerifierErrors>;
type VerifierPartialResult<T> = Result<T, ()>;
```
and then made every function that currently uses `VerifierResult` *either*:
 - use `VerifierPartialResult` and take an &mut `VerifierResults`, or
 - use `VerifierResult` and don't take an &mut `VerifierResults`
?

6A commented:
Okay, I modified all ``verify`` functions that weren't used "in groups" so that they all return `VerifierResult<()>` (aka `Result<(), VerifierErrors>`) and take not `&mut VerifierErrors`. All other functions now return `VerifierStepResult<()>` (aka `Result<(), ()>`).

I also added more details on how both of them should be used in their doc comments.

sunfishcode commented on lib/codegen/src/verifier/mod.rs:
```diff
@@ -82,23 +82,86 @@ pub use self::cssa::verify_cssa;
+    });
+}
+
+/// Faster syntax for calling `verify_*` functions.
```
"Shorthand" would convey the purpose here better than "faster". Also, a good one-line summary of this macro is that it adapts a function returning a `VerifierStepResult` to return a `VerifierResult`.

sunfishcode commented on lib/codegen/src/verifier/mod.rs:
````diff
@@ -82,23 +82,86 @@ pub use self::cssa::verify_cssa;
+/// ```rust,ignore
+/// // Instead of doing all this manually
+/// let mut errors = VerifierErrors::default();
+/// let result = verify_function(func, &mut errors);
````
I know this is only an example, but it should use `verify_context` or so which returns a `VerifierStepResult` to avoid confusion.

6A replied:
Good catch! I wrote the docs *before* changing the signature of `verify_function` so I went over this too quickly...

sunfishcode commented on lib/codegen/src/verifier/mod.rs:
```diff
@@ -82,23 +82,86 @@ pub use self::cssa::verify_cssa;
+/// }
+///
+/// // This macro allows you to write
+/// verify!(verify_func, func)
```
Update the name here too.

sunfishcode commented on lib/codegen/src/context.rs:
```diff
@@ -177,31 +177,43 @@ impl Context {
+            Ok(())
+        } else {
+            Err(errors)
+        }
```
Can this use `verify!`?

6A replied:
`context` is imported before `verifier` therefore the `verify!` macro doesn't exist yet when this is parsed, so either we change the order at which the modules are loaded or we keep this.

sunfishcode commented on lib/codegen/src/context.rs:
```diff
@@ -177,31 +177,43 @@ impl Context {
+            Ok(())
+        } else {
+            Err(errors)
+        }
```
Can this use `verify!`?

6A replied:
Ditto.

sunfishcode commented:
Looks good, thanks for seeing this through!
